File: solutions/elixir/space-age/1/lib/space_age.ex
defmodule SpaceAge do
  @earth_year_in_seconds 31_557_600
  @type planet ::
          :mercury
          | :venus
          | :earth
          | :mars
          | :jupiter
          | :saturn
          | :uranus
          | :neptune

  @doc """
  Return the number of years a person that has lived for 'seconds' seconds is
  aged on 'planet', or an error if 'planet' is not a planet.
  """
  @spec age_on(planet, pos_integer) :: {:ok, float} | {:error, String.t()}
  def age_on(planet, seconds) do
    case orbital_period_in_earth_years(planet) do
      {:ok, ratio} when is_float(ratio) -> {:ok, seconds / @earth_year_in_seconds / ratio}
      error -> error
    end
  end

  @spec orbital_period_in_earth_years(planet) :: {:ok, float} | {:error, String.t()}
  defp orbital_period_in_earth_years(:mercury), do: {:ok, 0.2408467}
  defp orbital_period_in_earth_years(:venus), do: {:ok, 0.61519726}
  defp orbital_period_in_earth_years(:earth), do: {:ok, 1.0}
  defp orbital_period_in_earth_years(:mars), do: {:ok, 1.8808158}
  defp orbital_period_in_earth_years(:jupiter), do: {:ok, 11.862615}
  defp orbital_period_in_earth_years(:saturn), do: {:ok, 29.447498}
  defp orbital_period_in_earth_years(:uranus), do: {:ok, 84.016846}
  defp orbital_period_in_earth_years(:neptune), do: {:ok, 164.79132}
  defp orbital_period_in_earth_years(_), do: {:error, "not a planet"}
end

File: solutions/java/space-age/1/src/main/java/SpaceAge.java
class SpaceAge {
  enum Planet {
    Mercury(0.2408467),
    Venus(0.61519726),
    Earth(1.0),
    Mars(1.8808158),
    Jupiter(11.862615),
    Saturn(29.447498),
    Uranus(84.016846),
    Neptune(164.79132);
    private final double orbit;

    Planet(double orbit) {
      this.orbit = orbit;
    }

    public double getOrbit() {
      return orbit;
    }
  }

  private final double seconds;

  SpaceAge(double seconds) {
    this.seconds = seconds;
  }

  double getSeconds() {
    return seconds;
  }

  double onEarth() {
    return onPlanet(Planet.Earth);
  }

  double onMercury() {
    return onPlanet(Planet.Mercury);
  }

  double onVenus() {
    return onPlanet(Planet.Venus);
  }

  double onMars() {
    return onPlanet(Planet.Mars);
  }

  double onJupiter() {
    return onPlanet(Planet.Jupiter);
  }

  double onSaturn() {
    return onPlanet(Planet.Saturn);
  }

  double onUranus() {
    return onPlanet(Planet.Uranus);
  }

  double onNeptune() {
    return onPlanet(Planet.Neptune);
  }

  double onPlanet(Planet planet) {
    return getSeconds() / 31557600 / planet.getOrbit();
  }
}
